File: appsets-windows-uwp/im/ImMessage.cs
using AA;
using AppSets.utils;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Media;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Windows.Devices.Sms;
using Windows.Foundation.Metadata;
using Windows.UI.Core;
using Windows.UI.ViewManagement;

#nullable enable
namespace AppSets.im
{
    public class Session
    {
        public ImObj imObj;
        public ConversationUiState conversationState;
    }
    public interface ImObj
    {
        public string id { get; set; }
        public string name { get; set; }
        public string avatar { get; set; }
        public bool isRelated { get; set; }

        public static MessageToInfo toToInfo(ImObj imObj) {
            if (imObj is ImSingle)
            {
                return new MessageToInfo()
                {
                    toType = "one2one",
                    id = imObj.id,
                    name = imObj.name,
                    iconUrl = imObj.avatar,
                    roles = (imObj as ImSingle).userRoles
                };
            }
            else if (imObj is ImGroup)
            {
                return new MessageToInfo()
                {
                    toType = "one2many",
                    id = imObj.id,
                    name = imObj.name,
                    iconUrl = imObj.avatar,
                };
            }
            else throw new Exception();
        }
    }

    public class ImGroup : ImObj
    {
        public string groupId;
        public string groupName;
        public string? groupAvatarUrl;
        public bool mIsRelated;

        public string id { get => groupId; set => groupId = value; }
        public string name { get => groupName; set => groupName = value; }
        public string avatar
        {
            get
            {
                if (string.IsNullOrEmpty(groupAvatarUrl))
                {
                    return "https://locahost";
                }
                return groupAvatarUrl;
            }
            set => groupAvatarUrl = value;
        }
        public bool isRelated { get => mIsRelated; set => mIsRelated = value; }
    }


    public class ImSingle : ImObj
    {
        public string uid;
        public string userName;
        public string? userAvatarUrl;
        public string? userRoles;
        public bool mIsRelated;

        public string id { get => uid; set => uid = value; }
        public string name { get => userName; set => userName = value; }
        public string avatar
        {
            get
            {
                if (string.IsNullOrEmpty(userAvatarUrl))
                {
                    return "https://localhost";
                }
                return userAvatarUrl;
            }
            set => userAvatarUrl = value;
        }
        public bool isRelated { get => mIsRelated; set => mIsRelated = value; }
    }

    public class ConversationUiState: BindableBase
    {
        public ObservableCollection<ImMessage> messages = new ObservableCollection<ImMessage>();

        private string? mLastImMessage = null;
        public string? latestImMessage
        {
            get {
                if (mLastImMessage==null) { 
                    return "无消息";
                }
                return mLastImMessage;
            }
            set {
                this.SetProperty(ref this.mLastImMessage, value);
            }
        }
        public ConversationUiState() {
            
        }

        

        internal void addMessage(ImMessage imMessage)
        {
            Debug.WriteLine("ConversationUiState addMessage Thread:" + Thread.CurrentThread.ManagedThreadId);
            messages.Add(imMessage);
            latestImMessage = imMessage.content;
        }
    }



    public class MessageFromInfo {
        public string id;
        public string? name;
        public string? avatarUrl;
        public string? roles;
    }

    public class MessageToInfo
    {
        public string id;
        public string? name;
        public string? iconUrl;
        public string? roles;
        public string toType;

        public bool isImgGroupMessage { get => toType == "one2many"; set => throw new Exception(); }
        public bool isImSingleMessage { get => toType == "one2one"; set => throw new Exception(); }
    }

    public interface ImMessage {
        public string id { get; set; }
        public string content { get; set; }
        public MessageFromInfo msgFromInfo { get; set; }
        public DateTime date { get; set; }
        public MessageToInfo msgToInfo { get; set; }
        public string? groupMessageTag { get; set; }
        public string contentType { get; set; }

        public static ImObj? parseFromImObj(ImMessage imMessage)
        {
            if (imMessage.msgToInfo.isImgGroupMessage)
            {
                return new ImGroup()
                {
                    id = imMessage.msgToInfo.id,
                    name = imMessage.msgToInfo.name,
                    avatar = imMessage.msgToInfo.iconUrl
                };
            }
            else if (imMessage.msgToInfo.isImSingleMessage)
            {
                return new ImSingle()
                {
                    id = imMessage.msgFromInfo.id,
                    name = imMessage.msgFromInfo.name,
                    avatar = imMessage.msgFromInfo.avatarUrl,
                    userRoles = imMessage.msgFromInfo.roles
                };
            }
            return null;
        }

        public static ImObj? parseToImObj(ImMessage imMessage)
        {
            if (imMessage.msgToInfo.isImgGroupMessage)
            {
                return new ImGroup()
                {
                    id = imMessage.msgToInfo.id,
                    name = imMessage.msgToInfo.name,
                    avatar = imMessage.msgToInfo.iconUrl
                };
            }
            else if (imMessage.msgToInfo.isImSingleMessage)
            {
                return new ImSingle()
                {
                    id = imMessage.msgToInfo.id,
                    name = imMessage.msgToInfo.name,
                    avatar = imMessage.msgToInfo.iconUrl,
                    userRoles = imMessage.msgToInfo.roles
                };
            }
            return null;
        }
    }

    public abstract class AbsImMessage:ImMessage
    {
        public abstract string id { get; set; }
        public abstract string content { get; set; }
        public abstract MessageFromInfo msgFromInfo { get; set; }
        public abstract DateTime date { get; set; }
        public abstract MessageToInfo msgToInfo { get; set; }
        public abstract string? groupMessageTag { get; set; }
        public abstract string contentType { get; set; }

        public Brush bgColor()
        {
            if (string.Equals(AccountProvider.Instance.UserInfo.Uid, msgFromInfo.id))
                {
                var color = new UISettings().GetColorValue(UIColorType.Accent);
          
                return new SolidColorBrush(color);
                }
                else
                {
                
                var color = new UISettings().GetColorValue(UIColorType.Accent);
                return new SolidColorBrush(color);
            }
        }

        public HorizontalAlignment msgAlignment()
        {
            if (string.Equals(AccountProvider.Instance.UserInfo.Uid, msgFromInfo.id))
            {
                return HorizontalAlignment.Right;
            }
            else
            {
                return HorizontalAlignment.Left;
            }
        }

        public CornerRadius bgCornerRadius()
        {
            if (string.Equals(AccountProvider.Instance.UserInfo.Uid, msgFromInfo.id))
            {
                return new CornerRadius(16, 2, 16, 16);
            }
            else
            {
                return new CornerRadius(2, 16, 16, 16);
            }
        }



    }
    public class ImText : AbsImMessage
    {
        public string msgId;
        public string text;
        public MessageFromInfo fromInfo;
        public DateTime timestamp;
        public MessageToInfo toInfo;
        public string? msgGroupTag;
        public ImText(string msgId, string text, MessageFromInfo fromInfo,
            DateTime timestamp, MessageToInfo toInfo, string? msgGroupTag)
        {
            this.msgId = msgId;
            this.text = text;
            this.fromInfo = fromInfo;
            this.timestamp = timestamp;
            this.toInfo = toInfo;
            this.msgGroupTag = msgGroupTag;

        }

        public override string id { get => msgId; set => msgId = value; }
        public override string content { get => text; set => text = value; }
        public override MessageFromInfo msgFromInfo { get => fromInfo; set => fromInfo = value; }
        public override DateTime date { get => timestamp; set => timestamp = value; }
        public override MessageToInfo msgToInfo { get => toInfo; set => toInfo = value; }
        public override string? groupMessageTag { get => msgGroupTag; set => msgGroupTag = value; }
        public override string contentType { get => "application/*"; set => throw new NotImplementedException(); }
    }


    class RabbitMqBrokerPropertyDesignType {
        public static string TYPE_TEXT = "im.content.text";
        public static string TYPE_IMAGE = "im.content.image";
        public static string TYPE_VIDEO = "im.content.video";
        public static string TYPE_MUSIC = "im.content.music";
        public static string TYPE_VOICE = "im.content.voice";
        public static string TYPE_LOCATION = "im.content.location";
        public static string TYPE_FILE = "im.content.file";
        public static string TYPE_HTML = "im.content.html";
        public static string TYPE_AD = "im.content.ad";
        public static string TYPE_SYSTEM = "im.content.system";
        public static string TYPE_CUSTOM = "im.content.custom.*";

        public static string getTypeByImMessage(ImMessage imMessage) {
            if (imMessage is ImText) { 
                return TYPE_TEXT;
            }
            return TYPE_CUSTOM;
        }
    }
}

File: appsets-android-native/appsets/src/main/kotlin/xcj/app/appsets/im/ImMessage.kt
package xcj.app.appsets.im

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import xcj.app.appsets.db.room.entity.FlatImMessage
import xcj.app.appsets.server.model.UserInfo
import java.util.Date

/**
 * @param groupMessageTag 当用户发送时选择了添加图片，然后又添加文字等一系列连串操作后，消息本身还是会进行串行发送，
 * 这一组有序列的消息的第一个消息将设置此字段为”(“， 序列最后的消息设置此字段为”)“.
 * @param dateStr 请根据规则提前设置此字段的值，尽量避免放到composition的时候做格式化
 */
sealed class ImMessage(
    val id: String,
    val content: String,
    val msgFromInfo: MessageFromInfo,
    val date: Date,
    val msgToInfo: MessageToInfo,
    val groupMessageTag: String?,
) {
    var notificationId: Int? = null
    var dateStr: String? = null
    val contentType: String = "application/*"
    fun contentIsLiteralType(): Boolean {
        return this is Text || this is HTML || this is Ad || this is Location
    }

    fun getFlatImMessage(): FlatImMessage {
        return FlatImMessage(
            id = id,
            content = content,
            uid = msgFromInfo.uid,
            name = msgFromInfo.name,
            avatarUrl = msgFromInfo.avatarUrl,
            roles = msgFromInfo.roles,
            timestamp = date,
            toId = msgToInfo.id,
            toName = msgToInfo.name,
            toIconUrl = msgToInfo.iconUrl,
            toType = msgToInfo.toType,
            groupMessageTag = groupMessageTag,
            messageType = RabbitMqBrokerPropertyDesignType.getTypeByImMessage(this)
        )
    }

    /**
     * 获取当前消息发送方是否是未关联的，发送方和自己有关联则返回null,没关联返回具体信息
     */
    fun getFromInfo(): UserInfo {
        return UserInfo.basicInfo(msgFromInfo.uid, msgFromInfo.name, msgFromInfo.avatarUrl)
    }

    fun contentByMyType(): String {
        return when (this) {
            is Text -> {
                content
            }

            is HTML -> {
                "${content} (website)"
            }

            is Ad -> {
                "${content} (advertisement)"
            }

            is Location -> {
                "${content} (location)"
            }

            is Music -> {
                "${musicJson.name} (music)"
            }

            is Video -> {
                "${videoJson.name} (video)"
            }

            is Voice -> {
                "${voiceJson.name} (voice)"
            }

            is File -> {
                "${fileJson.name} (file)"
            }

            is Image -> {
                "(image)"
            }

            is System -> {
                systemContentJson.content
            }
        }
    }

    data class HTML(
        val msgId: String, val url: String, val fromUserInfo: MessageFromInfo,
        val timestamp: Date, val toInfo: MessageToInfo, val msgGroupTag: String?
    ) : ImMessage(msgId, url, fromUserInfo, timestamp, toInfo, msgGroupTag)

    data class Text(
        val msgId: String, val text: String, val fromUserInfo: MessageFromInfo,
        val timestamp: Date, val toInfo: MessageToInfo, val msgGroupTag: String?
    ) : ImMessage(msgId, text, fromUserInfo, timestamp, toInfo, msgGroupTag)

    data class Image(
        val msgId: String,
        val imageJson: String,
        val imageRaw: String?,
        val fromUserInfo: MessageFromInfo,
        val timestamp: Date,
        val toInfo: MessageToInfo,
        val msgGroupTag: String?
    ) : ImMessage(msgId, imageJson, fromUserInfo, timestamp, toInfo, msgGroupTag)

    data class Voice(
        val msgId: String,
        val voiceJson: CommonURLJson.VoiceURLJson,
        val voiceRaw: String,
        val fromUserInfo: MessageFromInfo,
        val timestamp: Date,
        val toInfo: MessageToInfo,
        val msgGroupTag: String?
    ) : ImMessage(
        msgId,
        voiceRaw,
        fromUserInfo,
        timestamp,
        toInfo,
        msgGroupTag
    )

    data class Video(
        val msgId: String,
        val videoJson: CommonURLJson.VideoURLJson,
        val videoRaw: String,
        val fromUserInfo: MessageFromInfo,
        val timestamp: Date,
        val toInfo: MessageToInfo,
        val msgGroupTag: String?
    ) : ImMessage(
        msgId,
        videoRaw,
        fromUserInfo,
        timestamp,
        toInfo,
        msgGroupTag
    )


    data class Music(
        val msgId: String,
        val musicJson: CommonURLJson.MusicURLJson,
        val musicRaw: String,
        val fromUserInfo: MessageFromInfo,
        val timestamp: Date,
        val toInfo: MessageToInfo,
        val msgGroupTag: String?
    ) : ImMessage(
        msgId,
        musicRaw,
        fromUserInfo,
        timestamp,
        toInfo,
        msgGroupTag
    )

    data class Ad(
        val msgId: String,
        val adJson: String, val fromUserInfo: MessageFromInfo,
        val timestamp: Date, val toInfo: MessageToInfo, val msgGroupTag: String?
    ) : ImMessage(msgId, adJson, fromUserInfo, timestamp, toInfo, msgGroupTag)

    data class Location(
        val msgId: String,
        val geoJson: String, val fromUserInfo: MessageFromInfo,
        val timestamp: Date, val toInfo: MessageToInfo, val msgGroupTag: String?
    ) : ImMessage(msgId, geoJson, fromUserInfo, timestamp, toInfo, msgGroupTag)

    data class File(
        val msgId: String,
        val fileJson: CommonURLJson.FileURLJson,
        val fileRaw: String,
        val fromUserInfo: MessageFromInfo,
        val timestamp: Date,
        val toInfo: MessageToInfo,
        val contentType1: String,
        val msgGroupTag: String?,
    ) : ImMessage(
        msgId,
        fileRaw,
        fromUserInfo,
        timestamp,
        toInfo,
        msgGroupTag
    )


    data class System(
        val msgId: String,
        val systemContentJson: SystemContentJson,
        val fromUserInfo: MessageFromInfo,
        val timestamp: Date,
        val toInfo: MessageToInfo,
        val contentType1: String?,
        val msgGroupTag: String?
    ) : ImMessage(
        msgId,
        systemContentJson.content,
        fromUserInfo,
        timestamp,
        toInfo,
        msgGroupTag
    ) {
        val handling: MutableState<Boolean> = mutableStateOf(false)
    }
}

/**
 * 解析消息To信息
 */
fun ImMessage.parseToImObj(): ImObj? {
    if (msgToInfo.isImgGroupMessage) {
        return ImObj.ImGroup(msgToInfo.id, msgToInfo.name, msgToInfo.iconUrl)
    }
    if (msgToInfo.isImSingleMessage) {
        return ImObj.ImSingle(msgToInfo.id, msgToInfo.name, msgToInfo.iconUrl, msgToInfo.roles)
    }
    return null
}


/**
 * 解析消息from信息
 */
fun ImMessage.parseFromImObj(): ImObj? {
    if (msgToInfo.isImgGroupMessage) {
        return ImObj.ImGroup(msgToInfo.id, msgToInfo.name, msgToInfo.iconUrl)
    }
    if (msgToInfo.isImSingleMessage) {
        return ImObj.ImSingle(
            msgFromInfo.uid,
            msgFromInfo.name,
            msgFromInfo.avatarUrl,
            msgFromInfo.roles
        )
    }
    return null
}